### PR TITLE
Fix DCLLoss negative-term aggregation and add loop-based reference test

### PR DESCRIPTION
**Pull request by yvesyue**
`**Bugfix:** In `_loss`, we previously computed:

    L = -s_pos + logsumexp(sim_00) + logsumexp(sim_01)

But according to the DCL paper (Eq.6), it should be:

    L = -s_pos + logsumexp([sim_00, sim_01])

So we now concatenate `sim_00` and `sim_01` and compute a single `logsumexp(...)`.

**Test:** Added a loop-based reference implementation to verify correctness.

**Patch:** In edge case handling, changed:

    return 2 - out0.shape[0] * nn.functional.softmax(similarity, dim=0)

to:

    return torch.tensor((2 - out0.shape[0] * nn.functional.softmax(similarity, dim=0)).clamp(min=0.0))

This avoids negative loss values in rare cases and ensures numerical stability.
`

### CODE REVIEW

File: tests/loss/test_dcl_loss.py [hunk 1]
```diff
@@ -101,3 +101,48 @@ def test_dclwloss_forward(self, seed: int = 0) -> None:
...
+        # find positive index
+        pos_idx = i + N if i < N else i - N
+        sim_pos = torch.dot(z[i], z[pos_idx]) / temperature
+
+        # negatives are all except i and pos_idx
+        mask = torch.ones(2 * N, dtype=torch.bool)
+        mask[i] = False
+        mask[pos_idx] = False
+        sims = torch.matmul(z[i], z[mask].T) / temperature  # shape (2N-2,)
+        neg_term = torch.log(sims.exp().sum())
```
Comment (yutong-xiang-97) [inline tests/loss/test_dcl_loss.py:129]: Refrain from using `.exp()` as it may cause numerical instability. Consider using
```python
all_dims = tuple(range(sims.ndim))
scalar = torch.logsumexp(sims, dim=all_dims)
```

**Review by yutong-xiang-97 — CHANGES_REQUESTED** [review]

File: tests/loss/test_dcl_loss.py (hunk 1 above)
Comment (yvesyue) [reply tests/loss/test_dcl_loss.py:129]: thanks for pointing it out , I have changed here to : 
`sims = torch.matmul(z[i], z[mask].T) / temperature  # shape (2N-2,)
        all_dims = tuple(range(sims.ndim))
        neg_term = torch.logsumexp(sims, dim=all_dims)`

File: tests/loss/test_dcl_loss.py [hunk 2]
```diff
@@ -101,3 +101,48 @@ def test_dclwloss_forward(self, seed: int = 0) -> None:
...
+        losses.append(-sim_pos + neg_term)
+
+    return torch.stack(losses).mean()
+
+
+def test_dclloss_matches_reference() -> None:
```
[the remark this answers is not in the document]
Comment (yvesyue) [reply tests/loss/test_dcl_loss.py:135]: OK, got it.  I have moved the `def test_dclloss_matches_reference() -> None:` under the `class TestDCLLoss:` and use @pytest.mark.parametrize, locally mypy and pytest passed.

File: tests/loss/test_dcl_loss.py [hunk 3]
```diff
@@ -101,3 +101,48 @@ def test_dclwloss_forward(self, seed: int = 0) -> None:
...
+
+
+def dcl_reference(
+    z1: torch.Tensor, z2: torch.Tensor, temperature: float
+) -> torch.Tensor:
+    """
+    Loop-based reference:
+      L = avg_i[ -sim_pos(i) + log(sum_j exp(sim_neg(i,j))) ]
+    """
+    z1 = torch.nn.functional.normalize(z1, dim=1)
```
[the remark this answers is not in the document]
Comment (yvesyue) [reply tests/loss/test_dcl_loss.py:138]: Thanks for the clarification! I understand now — my use of variable names may have caused some confusion. I've updated the function to follow the naming convention (x1, x2 for unnormalized, z1, z2 for normalized).

File: tests/loss/test_dcl_loss.py [hunk 4]
```diff
@@ -101,3 +101,58 @@ def test_dclwloss_forward(self, seed: int = 0) -> None:
         loss1 = criterion(out1, out0)
         assert loss0 > 0
         assert loss0 == pytest.approx(loss1)
+
+    @pytest.mark.parametrize(
+        "batch_size, dim, temperature",
+        [
```
[the remark this answers is not in the document]
Comment (yvesyue) [reply tests/loss/test_dcl_loss.py:107]: I see now that separating the parameters improves clarity and granularity. I've updated the test to follow this format accordingly!
